Mikhail kasimov patch 1

MikhailKasimov opened the pull request:
Sorry, have missed one fix to add.

MikhailKasimov committed:
Update gost3411-2012.spec

MikhailKasimov committed:
Update gost3411-2012.spec

adegtyarev commented:
So what's the reason to rename main package's Group to "Development/Tools/Other"?

MikhailKasimov commented:
RPMLint errors on different groups for main package and its devel add-on. Group synchronization only.

adegtyarev commented:
Yeah, but I'd rather rename the Group of devel package to "Productivity/Security"

MikhailKasimov commented:
Two days ago I also thought in the same way, but faced with RPMLint error, that devel package should be in development group. And to put "Development/Tools/Other" both for main package and its devel add-on is the easiest way to have successfully built package for all RPM-based distros with no error messages. So, there are no conflicts.

MikhailKasimov committed:
Merge branch 'add-devel-rpm' into MikhailKasimov-patch-1

adegtyarev commented:
That way no package could be in any group other than in "Development", as long as both packages must share the same group and devel package should be in "Development". 

I have just synced both groups to "Productivity/Security" and see no problems with RPMLint:

```
# grep ^Group gost3411-2012.spec 
Group:          Productivity/Security
Group:          Productivity/Security

# rpmlint gost3411-2012.spec 
0 packages and 1 specfiles checked; 0 errors, 0 warnings.

# rpmlint --version
rpmlint version 1.8 Copyright (C) 1999-2007 Frederic Lepied, Mandriva
```

MikhailKasimov commented:
OK, just re-built RPM: https://build.opensuse.org/package/live_build_log/home:k_mikhail/GOST34.11-2012/openSUSE_Leap_42.3/x86_64

```
[   34s] +GOST34.11-2012-devel.x86_64: W: devel-package-with-non-devel-group Productivity/Security
```

That's why I moved ```Group: ``` from ```Productivity/Security``` to ```Development/Tools/Other``` for both of packages, which OK and gives no such crappy error-messages of RPMLint. Possibly, it is SUSE-specific issue just like a ```%dir %{_includedir}/gost34.11-2012``` for ```%files```, when RH-based distros are built normally instead of SUSE's ones.

MikhailKasimov commented:
UPD: So, SUSE specific: https://en.opensuse.org/openSUSE:Package_group_guidelines

```
Tools

Groups from this section contain tools useful for developing that are not connected to a particular programming language.

    Development/Tools/Building
    Development/Tools/Debuggers
    Development/Tools/Doc Generators
    Development/Tools/GUI Builders
    Development/Tools/IDE
    Development/Tools/Navigators
    Development/Tools/Other
    Development/Tools/Version Control
```

Sorry, but, please accept ```Group: Development/Tools/Other``` both for ```gost34.11-2012``` and ```gost34.11-2012-devel``` packages. Thanks!

adegtyarev commented:
The group name "Development/Tools/Other" for the __main__ package does not looks valid for me.

Following the guideline you linked I came up with:

- main package: *Productivity/Security* - Groups from this section are intended for security related stuff like virus scanners, safe password generators, utilities for encrypting, decrypting, signing data, and permission settings
- devel package: *Development/Libraries/C* - that group is intended for packages that allow developing with a library. They are primarily sorted by programming language.

Committed to branch 'add-devel-rpm', please review.

MikhailKasimov commented:
> Committed to branch 'add-devel-rpm', please review.

https://build.opensuse.org/package/live_build_log/home:k_mikhail/GOST34.11-2012/openSUSE_Leap_42.3/x86_64

And again:

```
[   30s] -GOST34.11-2012-devel.x86_64: W: devel-package-with-non-devel-group Productivity/Security
```

MikhailKasimov commented:
> The group name "Development/Tools/Other" for the main package does not looks valid for me.

Me too. But I have to accept some items, even if they look dumb. Before making ```Development/Tools/Other``` for both of packages, I had to try out a couple of variations to say at last 'f..ck that!' and revert to  ```Development/Tools/Other``` to have errors-free package.

**UPD**: Just for example: here is ```wxhexeditor```, which is just hex-editor, nothing more... But it also has  ```Development/Tools/Other```.  Here is its ```spec``` to compare: https://build.opensuse.org/package/view_file/editors/wxhexeditor/wxhexeditor.spec?expand=1 See ```Group:``` tag. I suppose, reason is the same to ours...

MikhailKasimov commented:
Finally success: ``` Development/Libraries/C and C++```, not ``` Development/Libraries/C```

https://build.opensuse.org/package/live_build_log/home:k_mikhail/GOST34.11-2012/openSUSE_Leap_42.3/x86_64

```
[   69s] RPMLINT report:
[   69s] ===============
[   69s] 3 packages and 0 specfiles checked; 0 errors, 0 warnings.

```

MikhailKasimov committed:
Update gost3411-2012.spec

MikhailKasimov committed:
Merge branch 'add-devel-rpm' into MikhailKasimov-patch-1

MikhailKasimov committed:
Update gost3411-2012.spec

MikhailKasimov committed:
Update gost3411-2012.spec

MikhailKasimov committed:
Pre-release gost34.11-2012.spec for devel-package

Final pre-release version of new ```gost34.11-2012.spec``` file to build ```gost34.11-2012``` and ```gost34.11-2012-devel``` packages.

Changelog for devel-package:

1. Fixed ```Group: Development/Libraries/C and C++``` (was incorrect ```Group: Development/Libraries/C```)

2. Fixed ```%description``` fields due to author's text.

3. ```Requires:  %{name}-devel = %{version}``` and ```Requires:  %{name} = %{version}``` for ```devel-package``` are left actual. Control of version's matching is strongly desirable because of avoiding possible mismatches on versions between main and devel packages. Hope strongly, this won't be ignored and will be accepted.

adegtyarev commented:
I still cannot understand & accept why a devel-package would require a main package? We can install any of both packages regardless of installing a second one.

MikhailKasimov commented:
Yes, in general way, these two packages can be installed separately. But devel-package (in particular, in our case) is only set of header files for ```/usr/include/``` folder and there's no sense of them if main package is not installed. They are orphaned without main package.

Here is, just for an example, snippet of spec-file for ```python3```: https://build.opensuse.org/package/view_file/devel:languages:python:Factory/python3/python3-base.spec?expand=1

```
%package -n python3-devel
Summary:        Include Files and Libraries Mandatory for Building Python Modules
Group:          Development/Languages/Python
Requires:       %{name} = %{version}

%description -n python3-devel
The Python programming language's interpreter can be extended with
dynamically loaded extensions and can be embedded in other programs.

This package contains header files, a static library, and development
tools for building Python modules, extending the Python interpreter or
embedding Python in applications.

This also includes the Python distutils, which were in the Python
package up to version 2.2.2.

%package -n python3-testsuite
Summary:        Unit tests for Python and its standard library
Group:          Development/Languages/Python
Requires:       python3 = %{version}
Requires:       python3-tk = %{version}

```

And ```Requires:``` on main package (``` Requires:       python3 = %{version}```) is normal and adequate practice. Why not to use such in our case? And, again, this avoids mess on versions between main and devel packages. Personally, I want to avoid such mess and to be calm on that. Mess = vulnerability (in generic case). Something like this...

adegtyarev commented:
> and there's no sense of them if main package is not installed.

No, that's wrong.

You install devel-package to have the [GOST 34.11-2012 engine](https://github.com/adegtyarev/streebog#api) available to build **your** application:

```
# rpm -i /usr/src/packages/RPMS/x86_64/GOST34.11-2012-devel-*.rpm

# gcc -I. -I/usr/include/gost3411-2012 your-cool-app-with-gost-digest.c
```

The CLI tool in the main package is just that kind of application, to be exact. Check the docs for the [example of usage](https://github.com/adegtyarev/streebog#example-of-usage).

MikhailKasimov commented:
Yes, and you are right here. The main problem here is -- both of us are right: 1) to give external developers possibility to use ```gost34.11-2012``` header files in their own applications and 2) to keep versions between ```gost34.11-2012``` main and devel packages synchronized. What shall we do here?

MikhailKasimov commented:
**UPD**: Found a package, where devel is not depended on its main package: https://build.opensuse.org/package/view_file/openSUSE:Leap:15.0/4ti2/4ti2.spec?expand=1

```
Name: 4ti2
...
%package devel
Summary:        Development files for 4ti2
Group:          Development/Libraries/C and C++
Requires:       lib4ti2-0 = %version
Requires:       libzsolve0 = %version
```

OK, let's have your variation in release, although I'm still not sure, this is correct.

MikhailKasimov committed:
Merge branch 'add-devel-rpm' into MikhailKasimov-patch-1

MikhailKasimov commented:
Closed on: https://github.com/adegtyarev/streebog/commit/e34b5d99b55ad08894143f6b068c7ca300b7e21e

adegtyarev commented:
> 2) to keep versions between gost34.11-2012 main and devel packages synchronized.

You are referring to a Python example but nope, that not the case here. Both packages does **not** have a mutual dependency on themselves.

Merged via #17 and published.  Thanks for making Streebog better!

MikhailKasimov commented:
No problem at all.  Thanks to you for being patient on my activity! :+1:  :grinning:

P.S. Package with new spec is already uploaded on OBS.
